toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,269 @@
+```
+  "app_ssh_ws_endpoint": "wss://ssh.cf-app.com"
+```
+
+- The URL presented in the field MUST be `wss://` + the route as announced by route-registrar for SSH proxy (see [CF Deployment](#cf-deployment))
+
+#### SSH Proxy
+
+- Diego SSH Proxy SHOULD provide the option to [launch a separate WebSocket server](https://github.com/cloudfoundry/diego-ssh/blob/main/cmd/ssh-proxy/main.go#L65) besides the regular SSH server
+- SSH Proxy SHOULD provide a separate handler for unwrapping WebSocket connections:
+```
+	sshWsProxy := wsProxy.New(logger, proxySSHServerConfig, metronClient, tlsConfig)
+	wsServer := server.NewServer(logger, sshProxyConfig.Address, websocket.Handler(sshWsProxy), time.Duration(sshProxyConfig.IdleConnectionTimeout))
+```
+Changes in [ssh-proxy/main.go](https://github.com/cloudfoundry/diego-ssh/blob/main/cmd/ssh-proxy/main.go#L65C1-L67C1)
+
+```
+func (p *WsProxy) HandleConnection(wsConn *websocket.Conn) {
+	logger := p.logger.Session("handle-ws-connection")
+	defer wsConn.Close()
+
+    // Here we start a SSH handshake inside the WebSocket tunnel
+	serverConn, serverChannels, serverRequests, err := ssh.NewServerConn(wsConn, p.serverConfig)
+
+(...)
+```
+Changes in [proxy.go](https://github.com/cloudfoundry/diego-ssh/blob/main/proxy/proxy.go#L67)
+
+**Remark:**
+(these changes are not meant as verbatim code; they only illustrate areas that would need to change)
+
+
+- SSH Proxy MUST retain the legacy SSH server on port 2222 for backwards compatibiliy
+- Diego Release SHOULD provide [a spec](https://github.com/cloudfoundry/diego-release/blob/develop/jobs/ssh_proxy/spec#L53) to define a `diego.ssh_proxy.listen_ws_addr` property with a default of `0.0.0.0:443` or `0.0.0.0:9443`
+
+#### CF Deployment
+
+- CF Deployment [deploys SSH Proxy on Scheduler](https://github.com/cloudfoundry/cf-deployment/blob/main/cf-deployment.yml#L1329) but there is no route to it at the moment.
+- CF Deployment SHOULD deploy [route-registrar](https://github.com/cloudfoundry/routing-release/tree/develop/jobs/route_registrar) alongside SSH Proxy and configure it to announce a TLS route at Gorouter:
+```
+  - name: route_registrar
+    properties:
+      nats:
+        tls:
+          enabled: true
+          client_cert: "((nats_client_cert.certificate))"
+          client_key: "((nats_client_cert.private_key))"
+      route_registrar:
+        routes:
+        - name: ssh-ws-proxy
+          port: 443     # or 9443 if preferred
+          tls_port: 443 # or 9443 if preferred
+          registration_interval: 20s
+          server_cert_domain_san: ssh.((system_domain))
+          uris:
+          - ssh.((system_domain))       # e.g. ssh.cf-app.com, announced by CAPI as wss://ssh.cf-app.com
+    release: routing
+```
+
+
+### Changes needed for regular SSH, SCP and SFTP
+
+- Getting the ssh token via `cf ssh-code` remains unchanged as it connects to UAA.
+- Regular ssh, scp and sftp commands SHOULD support a `-o ProxyCommand` option.
```

Comment:
> I would even go a bit further and say that the CF CLI must add support for cf proxy-ssh or whatever.

I had the same thought initially, but I figured people who use regular ssh instead of cf ssh usually don't have the CF CLI installed, so a `cf proxy-ssh` command wouldn't make much sense. But I am open for putting it in two places: 
- CF CLI
- separate binary

Reply:
For SCP you can't use the CF CLI. In such cases most users probably use the cf cli to generate the token which is then (manually) passed on to scp. Having the proxy built-in to the CLI would help in that case as it avoids the need to install a separate binary.